types/wlr_presentation_time.c

```diff
@@ -63,18 +51,38 @@ static void feedback_handle_surface_commit(struct wl_listener *listener,
 	if (feedback->committed) {
 		if (!feedback->sampled) {
 			// The content update has been superseded
-			feedback_send_discarded(feedback);
+			wlr_presentation_feedback_destroy(feedback);
 		}
 	} else {
 		feedback->committed = true;
 	}
 }
 
+static void feedback_unset_surface(struct wlr_presentation_feedback *feedback) {
+	if (feedback->surface == NULL) {
+		return;
+	}
+
+	feedback->surface = NULL;
+	wl_list_remove(&feedback->surface_commit.link);
+	wl_list_remove(&feedback->surface_destroy.link);
+}
+
 static void feedback_handle_surface_destroy(struct wl_listener *listener,
 		void *data) {
 	struct wlr_presentation_feedback *feedback =
 		wl_container_of(listener, feedback, surface_destroy);
-	feedback_send_discarded(feedback);
+	if (feedback->sampled) {
+		// The compositor might have a handle on this feedback
+		struct wl_resource *resource, *tmp;
+		wl_resource_for_each_safe(resource, tmp, &feedback->resources) {
+			feedback_resource_send_discarded(resource);
```

Comment:
Since this feedback was already sampled, isn't it possible for the compositor to present the sampled contents even after the surface has been destroyed, necessitating a `presented` rather than `discarded`?

Reply:
Per the protocol, `discarded` should be sent if the surface is destroyed:

> or, the user did not see the content update because it was superseded or its surface destroyed, and the content update is discarded.

Anyway, if the surface is destroyed, it doesn't really matter, we just want to cleanup the feedback objects properly.

Reply:
> Per the protocol, `discarded` should be sent if the surface is destroyed:
> 
> > or, the user did not see the content update because it was superseded or its surface destroyed, and the content update is discarded.

This reads to me as "the user did not see the content update because \<for example\> its surface was destroyed \<before the compositor had a chance to obtain its contents\>", but I might be mistaken.

> Anyway, if the surface is destroyed, it doesn't really matter, we just want to cleanup the feedback objects properly.

Alright; if it's fine by wlroots internals then it works.

Reply:
Ah, I see, you're right. It's not really more work to send `presented`, so I've just updated the PR to let the compositor do that instead.